### ISSUE TITLE
false positive: question_mark when returning custom Error type 

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#7859]
Lint name: question_mark

I tried this code:

```rust
    let mut cwd: PathBuf = if let Ok(cwd) = env::current_dir() {
        cwd
    } else {
        return Err(Error::NoCWD);
    };
```

I expected to see this happen:
There should be no warning.
`Error::NoCWD` is my own error type (an enum) and it does not impl `std::convert::From<std::io::Error>` or anything like that. So I can't use `?` on it directly.

Instead, this happened: 

````
warning: this if-let-else may be rewritten with the `?` operator
  --> src/commands/local.rs:53:28
   |
53 |       let mut cwd: PathBuf = if let Ok(cwd) = env::current_dir() {
   |  ____________________________^
54 | |         cwd
55 | |     } else {
56 | |         return Err(Error::NoCWD);
57 | |     };
   | |_____^ help: replace it with: `env::current_dir()?`
   |
note: the lint level is defined here
````
### Meta

`clippy 0.1.58 (68a698b 2021-10-21)`

```
rustc 1.58.0-nightly (68a698baf 2021-10-21)
binary: rustc
commit-hash: 68a698baf6bfc61d85ce6e25122a092c60c7f21a
commit-date: 2021-10-21
host: x86_64-unknown-linux-gnu
release: 1.58.0-nightly
LLVM version: 13.0.0
````

**Rust version (`rustc -Vv`):**
```
rustc 1.46.0-nightly (f455e46ea 2020-06-20)
binary: rustc
commit-hash: f455e46eae1a227d735091091144601b467e1565
commit-date: 2020-06-20
host: x86_64-unknown-linux-gnu
release: 1.46.0-nightly
LLVM version: 10.0
```

@ rustbot label +I-suggestion-causes-error

**Comment by matthiaskrgr** [comment]
```
error[E0277]: `?` couldn't convert the error to `library::Error`
   --> src/commands/local.rs:53:46
    |
51  | pub(crate) fn get_manifest() -> Result<PathBuf, Error> {
    |                                 ---------------------- expected `library::Error` because of this
52  |     // get the cwd
53  |     let mut cwd: PathBuf = env::current_dir()?;
    |                                              ^ the trait `From<std::io::Error>` is not implemented for `library::Error`
    |
    = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
    = note: required because of the requirements on the impl of `FromResidual<Result<Infallible, std::io::Error>>` for `Result<PathBuf, library::Error>`
note: required by `from_residual`
   --> /home/matthias/.rustup/toolchains/master/lib/rustlib/src/rust/library/core/src/ops/try_trait.rs:339:5
    |
339 |     fn from_residual(residual: R) -> Self;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0277`.
```
